src/io/import/importDataSources.ts

```diff
@@ -58,72 +35,138 @@ const unhandledResource: ImportHandler = () => {
 };
 
 export async function importDataSources(dataSources: DataSource[]) {
-  const middleware: Array<ImportHandler> = [
+  const importContext = {
+    fetchFileCache: new Map<string, File>(),
+    dicomDataSources: [] as DataSourceWithFile[],
+  };
+
+  const middleware = [
     // updating the file type should be first in the pipeline
     updateFileMimeType,
-    // handleConfig must before restoreStateFile for label props to be applied to deserialized tools
-    handleConfig,
+    // before extractArchive as .zip extension is part of state file check
     restoreStateFile,
     handleRemoteManifest,
     handleGoogleCloudStorage,
     handleAmazonS3,
     downloadUrl,
     extractArchiveTargetFromCache,
     extractArchive,
+    handleConfig, // collect to apply later
     // should be before importSingleFile, since DICOM is more specific
     handleDicomFile,
     importSingleFile,
     // catch any unhandled resource
     unhandledResource,
   ];
+  const loader = new Pipeline(middleware);
 
-  const importContext = {
-    fetchFileCache: new Map<string, File>(),
-    dicomDataSources: [] as DataSourceWithFile[],
-  };
-
-  const pipeline = new Pipeline(middleware);
   const results = await Promise.all(
-    dataSources.map((r) => pipeline.execute(r, importContext))
+    dataSources.map((r) => loader.execute(r, importContext))
   );
 
-  if (!importContext.dicomDataSources.length) {
-    return results;
-  }
+  // Apply config.JSONs
+  const configResult = await (async () => {
+    try {
+      results
+        .flatMap((pipelineResult) =>
+          pipelineResult.ok ? pipelineResult.data : []
+        )
+        .map((result) => result.config)
+        .filter(nonNullable)
+        .forEach(applyConfig);
+      return {
+        ok: true as const,
+        data: [],
+      };
+    } catch (err) {
+      return {
+        ok: false as const,
+        errors: [
+          {
+            message: toMeaningfulErrorString(err),
+            cause: err,
+            inputDataStackTrace: [],
+          },
+        ],
+      };
+    }
+  })();
 
   // handle DICOM loading
-
   const dicomDataSource: DataSource = {
     dicomSrc: {
       sources: importContext.dicomDataSources,
     },
   };
-  const dicomResult: PipelineResult<DataSource, ImportResult> = {
-    ok: true,
-    data: [],
-    errors: [],
-  };
+  const dicomResult = await (async () => {
+    try {
```

Comment:
Let's split out these DICOM loading and config loading IIFEs into separate functions in this file for readability.